plugins/modules/ec2_metadata_facts.py

```diff
@@ -525,7 +538,29 @@ def fix_invalid_varnames(self, data):
 
         return new_data
 
+    def fetch_session_token(self, uri_token):
+        """Used to get a session token for IMDSv2"""
+        headers = {'X-aws-ec2-metadata-token-ttl-seconds': '60'}
+        response, info = fetch_url(self.module, uri_token, method='PUT', headers=headers, force=True)
+
+        if info.get('status') == 403:
+            self.module.fail_json(msg='Failed to retrieve metadata token from AWS: {0}'.format(info['msg']), response=info)
+        elif info.get('status') not in (200, 404):
+            time.sleep(3)
+            # request went bad, retry once then raise
+            self.module.warn('Retrying query to metadata service. First attempt failed: {0}'.format(info['msg']))
+            response, info = fetch_url(self.module, uri_token, method='PUT', headers=headers, force=True)
+            if info.get('status') != 200:
+                # fail out now
+                self.module.fail_json(msg='Failed to retrieve metadata token from AWS: {0}'.format(info['msg']), response=info)
```

Comment:
My concern is around the behaviour here.  A 404 message from a not-Amazon-service resulting in not being able to get a session token should be handled gracefully.

I'm good with the idea of defaulting to IMDSv2 being enabled by default, and it's probably even the right thing, but I'd like for the code to cleanly fall back to v1 when v2 isn't supported.  Having a parameter to *only* support v2 and fail here would be reasonable, but by default it should return None and continue without a token.

I think it would also be reasonable for the default behaviour to change after a deprecation period, but we should be issuing a warning for 18 months or so, such that someone who knows they need v1 behaviour can explicitly update their defaults (they can also then push vendors to add v2 support.

Reply:
Following the logic through again, the issue I was worried about would only have happened if we got something like a 500 followed by a 404.